### PR TITLE
source_location

### PR DESCRIPTION
**Pull request by andycroll**

### CODE REVIEW

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+category: ruby
+image:
+  base: "2023/find-definitions-of-rails-methods-using-source_location"
+  alt: "Landsat maps in black and white"
+  credit: "USGS"
+  source: "https://unsplash.com/photos/white-and-black-abstract-painting-PuLsDCBbyBM"
+
+---  
+
+As a dynamic language it's important that Ruby provides many excellent debugging and introspection features.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Given it's a dynamic language, it's important that Ruby comes with several excellent debugging and introspection features out of the box.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+Finding the place a method or block of code is defined and reading its source code is essential for effective debugging and code comprehension. In Ruby, `source_location`, provides a powerful tool for retrieving the file and line number where a particular method or block is defined.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Finding the exact place a method or block of code is defined and being able to read the related source code is essential for effective debugging and code comprehension. In Ruby, the `source_location` method, provides a powerful tool for retrieving the file and line number where a particular method or block is defined.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+#=> ["/.../activesupport-7.0.8/lib/active_support/core_ext/string/inflections.rb", 60]
+```
+
+Then open from the command line:
+
+```shell
+bundle open activesupport
+```
+
+Which opens the named gem in your editor of choice for you to explore.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
...which opens the named gem in your editor of choice for you to explore.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+
+## Why?
+
+Use of `source_location` is invaluable when you're new to a application, re-exploring unfamiliar code or  trying to understand which gem is providing functionality you're using.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Use of `source_location` is invaluable when you're new to an application, re-exploring unfamiliar code, or  trying to understand which gem is providing the functionality you're using.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+Reading source code is a great way to learn. Reading battle-tested code like Rails, or other gems, even more so.
+
+Thanks to authors of the framework it even works for some of the meta-programmed "generated methods" in Rails. Methods on Active Record [associations that are generated](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/associations/builder/association.rb#L103) using `class_eval` pass special syntax to enable the `source_location` lookup to still work (or you'd just see `(eval)` as the first result.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]: Should the hyperlink include the 'class_eval' part?

Comment (nodunayo) [reply _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]: I think this paragraph is a little jumbled, especially towards the end?

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+Then open from the command line:
+
+```shell
+bundle open activerecord
+```
+
+## Why not?
+
+If you experiment you might find that this doesn't work.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
While you're experimenting, you might find that this doesn't always work.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+```ruby
+"Cars".method(:upcase).source_location
+#=> nil
+```
+
+Methods included as part of "core" Ruby are often implemented in C, and their definitions are not directly accessible from Ruby code. Therefore, calling `source_location` on a core method will typically return `nil`.
+
+The `source_location` method only works for methods defined in Ruby or in Ruby gems where the source code is available. You might also have trouble if the method in question is defined in a C extension, or even when if you do get to the source code, it'll just be a call out to C code!
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]: Ending here seems a little jumbled too!

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+Also while `source_location` can be invaluable during development and debugging, it's not generally suitable for regular use in production code.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Also, while `source_location` can be invaluable during development and debugging, it's not generally suitable for regular use in production code. 
```

**Review by nodunayo — DISMISSED** [review]

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md (hunk 5 above)
Comment (andycroll) [reply _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Thanks to authors of the framework `source_location` works for "magic" meta-programmed methods in Rails. Methods on Active Record [associations that are generated using `class_eval`](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/associations/builder/association.rb#L103) pass special syntax to enable the lookup to still work. If it weren't for this when you called `source_location` you'd just see `(eval)` as the first result.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md (hunk 7 above)
Comment (andycroll) [reply _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
The `source_location` method only works for methods defined in gems where the source code is in Ruby. It won't work for methods that use a C extension (where Ruby code calls out to C).
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  base: "2023/find-definitions-of-rails-methods-using-source_location"
+  alt: "Landsat maps in black and white"
+  credit: "USGS"
+  source: "https://unsplash.com/photos/white-and-black-abstract-painting-PuLsDCBbyBM"
+
+---  
+
+Given it's a dynamic language, it's important that Ruby comes with several excellent debugging and introspection features out of the box.
+
+Finding the exact place a method or block of code is defined and being able to read the related source code is essential for effective debugging and code comprehension. In Ruby, the `source_location` method, provides a powerful tool for retrieving the file and line number where a particular method or block is defined.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Finding the exact place a method or block of code is defined and being able to read the related source code is essential for effective debugging and code comprehension. In Ruby, the `source_location` method, provides a powerful tool for retrieving the file and line number for where a particular method or block is defined.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+```shell
+bundle open activesupport
+```
+
+...which opens the named gem in your editor of choice for you to explore.
+
+
+## Why?
+
+Use of `source_location` is invaluable when you're new to an application, re-exploring unfamiliar code, or  trying to understand which gem is providing the functionality you're using.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Use of `source_location` is invaluable when you're new to an application, re-exploring unfamiliar code, or trying to understand which gem is providing the functionality you're using.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+Reading source code is a great way to learn. Reading battle-tested code like Rails, or other gems, even more so.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Reading source code is a great way to learn. Reading battle-tested code like that of Rails itself, or other gems, even more so.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+Thanks to authors of the framework `source_location` works for "magic" meta-programmed methods in Rails. Methods on Active Record [associations that are generated using `class_eval`](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/associations/builder/association.rb#L103) pass special syntax to enable the lookup to still work. If it weren't for this when you called `source_location` you'd just see `(eval)` as the first result.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Thanks to the authors of the framework, `source_location` also works for "magic" meta-programmed methods in Rails. Methods on Active Record [associations that are generated using `class_eval`](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/associations/builder/association.rb#L103) pass special syntax to enable the lookup to still work. If it weren't for this, when you called `source_location` on these methods, you'd always just see `(eval)` as the first result.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  has_many :seats
+
+  # ...
+end
+
+Car.first.method(:seats).source_location
+#=> [".../activerecord-7.0.8/lib/active_record/associations/builder/association.rb", 103]
+```
+
+Then open from the command line:
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md:56]: Do you need to show this part of the second example?

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+```shell
+bundle open activerecord
+```
+
+## Why not?
+
+While you're experimenting, you might find that this doesn't always work.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
While you're experimenting, you might find that this method doesn't always provide a helpful result.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+```ruby
+"Cars".method(:upcase).source_location
+#=> nil
+```
+
+Methods included as part of "core" Ruby are often implemented in C, and their definitions are not directly accessible from Ruby code. Therefore, calling `source_location` on a core method will typically return `nil`.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]: Maybe this is better as the first paragraph of this section, then the example.

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+The `source_location` method only works for methods defined in gems where the source code is in Ruby. It won't work for methods that use a C extension (where Ruby code calls out to C).
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
It also won't work for methods that use a C extension (where Ruby code calls out to C). `source_location` only works for methods defined in gems where the source code is in Ruby. 
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+Also, while `source_location` can be invaluable during development and debugging, it's not generally suitable for regular use in production code.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]: When would someone even think to include this in their production code though?

**Review by nodunayo — DISMISSED** [review]

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md (hunk 13 above)
Comment (andycroll) [reply _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md:56]: I think so. Folks don't know how to open gem source code.

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md (hunk 17 above)
Comment (andycroll) [reply _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]: By accident. Seen it done!

Comment (andycroll) [reply _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Also, while `source_location` can be invaluable during development and debugging, don't accidentally include in production code!
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md (hunk 15 above)
Comment (andycroll) [reply _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]: Done.

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+  base: "2023/find-definitions-of-rails-methods-using-source_location"
+  alt: "Landsat maps in black and white"
+  credit: "USGS"
+  source: "https://unsplash.com/photos/white-and-black-abstract-painting-PuLsDCBbyBM"
+
+---  
+
+Given it's a dynamic language, it's important that Ruby comes with several excellent debugging and introspection features out of the box.
+
+Finding the exact place a method or block of code is defined and being able to read the related source code is essential for effective debugging and code comprehension. In Ruby, the `source_location` method, provides a powerful tool for retrieving the file and line number for where a particular method or block is defined.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
Finding the exact place a method or block of code is defined, and being able to read the related source code, is essential for effective debugging and code comprehension. In Ruby, the `source_location` method provides a powerful tool for retrieving the file and line number for where a particular method or block is defined.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+#=> ["/.../activesupport-7.0.8/lib/active_support/core_ext/string/inflections.rb", 60]
+```
+
+Then open from the command line:
+
+```shell
+bundle open activesupport
+```
+
+...which opens the source code of the gem in your editor of choice for you to explore.
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
...which opens the source code of the gem in your editor of choice.
```

File: _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+While you're experimenting, you might find `source_location` doesn't always provide a helpful result. Methods included as part of "core" Ruby are often implemented in C, and their definitions are not directly accessible from Ruby code. Therefore, calling `source_location` on a core method will typically return `nil`.
+
+```ruby
+"Cars".method(:upcase).source_location
+#=> nil
+```
+
+It also won't work for methods that use a C extension (where Ruby code calls out to C). `source_location` only works for methods defined in gems where the source code is in Ruby. 
+
+Also, while `source_location` can be invaluable during development and debugging, don't accidentally include in production code!
```
Comment (nodunayo) [inline _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]:
```suggestion
And while `source_location` can be invaluable during development and debugging, don't accidentally include it in production code!
```

Comment (nodunayo) [reply _posts/2023/2023-12-04-find-definitions-of-rails-methods-using-source_location.md]: Much better way of stating this!!

**Pull request merged**